File: Components/DevExpressVCL/Demos/ExpressGauge Control/Delphi/WorldTime/uWorldTime.pas
unit uWorldTime;

interface

uses
  Forms, StdCtrls, Classes, Controls, ExtCtrls, cxLookAndFeels, cxLookAndFeelPainters, cxGraphics, cxClasses, cxControls,
  dxGaugeCustomScale, dxGaugeQuantitativeScale, dxGaugeCircularScale, dxGaugeControl, dxGaugeDigitalScale, cxSplitter,
  cxContainer, cxEdit, cxLabel;

type
  TTimeInfo = record
    Hour: Word;
    Min: Word;
    Sec: Word;
  end;

  TfrmWorldTime = class(TForm)
    Timer1: TTimer;
    pnlClocks: TPanel;
    pnlWashingtonClock: TPanel;
    gcWashingtonTime: TdxGaugeControl;
    gsWashingtonTimeBackgroundLayer: TdxGaugeCircularScale;
    gsWashingtonTimeHourNeedle: TdxGaugeCircularScale;
    gsWashingtonTimeMinuteNeedle: TdxGaugeCircularScale;
    gsWashingtonTimeSecondNeedle: TdxGaugeCircularScale;
    pnlParisClock: TPanel;
    pnlMoscowClock: TPanel;
    gcParisTime: TdxGaugeControl;
    dxGaugeCircularScale1: TdxGaugeCircularScale;
    dxGaugeCircularScale2: TdxGaugeCircularScale;
    dxGaugeCircularScale3: TdxGaugeCircularScale;
    dxGaugeCircularScale4: TdxGaugeCircularScale;
    gcMoscowTime: TdxGaugeControl;
    dxGaugeCircularScale5: TdxGaugeCircularScale;
    dxGaugeCircularScale6: TdxGaugeCircularScale;
    dxGaugeCircularScale7: TdxGaugeCircularScale;
    dxGaugeCircularScale8: TdxGaugeCircularScale;
    gcLocalTime: TdxGaugeControl;
    gsDigital: TdxGaugeDigitalScale;
    pnlLocalTime: TPanel;
    cxSplitter1: TcxSplitter;
    cxLabel1: TcxLabel;
    cxLabel3: TcxLabel;
    cxLabel2: TcxLabel;
    lbDescription: TcxLabel;
    cxLookAndFeelController1: TcxLookAndFeelController;
    procedure FormCreate(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure Timer1Timer(Sender: TObject);
    procedure pnlClocksResize(Sender: TObject);
  protected
    function GetTimeInfo(ATimeZone: Integer): TTimeInfo;
    procedure UpdateClock(AGaugeControl: TdxGaugeControl; ATimeInfo: TTimeInfo);
    procedure UpdateClocks;
  end;

var
  frmWorldTime: TfrmWorldTime;

implementation

{$R *.dfm}

uses
  Types, Windows, SysUtils, cxGeometry, cxDateUtils;

const
  Offset = 5;

{ TfrmWorldTime }

function TfrmWorldTime.GetTimeInfo(ATimeZone: Integer): TTimeInfo;
var
  AMilli : Word;
  ASystemTime: TSystemTime;
begin
  GetSystemTime(ASystemTime);
  DecodeTime(SystemTimeToDateTime(ASystemTime), Result.Hour, Result.Min, Result.Sec, AMilli);
  Result.Hour := (Result.Hour + ATimeZone) mod 12;
end;

procedure TfrmWorldTime.UpdateClock(AGaugeControl: TdxGaugeControl; ATimeInfo: TTimeInfo);
begin
  AGaugeControl.BeginUpdate;
  TdxGaugeCircularScale(AGaugeControl.Scales[3]).Value := ATimeInfo.Min;
  TdxGaugeCircularScale(AGaugeControl.Scales[2]).Value := ATimeInfo.Hour + ATimeInfo.Min / 60;
  TdxGaugeCircularScale(AGaugeControl.Scales[1]).Value := ATimeInfo.Sec;
  AGaugeControl.EndUpdate;
end;

procedure TfrmWorldTime.UpdateClocks;
begin
  UpdateClock(gcWashingtonTime, GetTimeInfo(-4));
  UpdateClock(gcParisTime, GetTimeInfo(1));
  UpdateClock(gcMoscowTime, GetTimeInfo(3));
  gsDigital.Value := cxTimeToStr(Now, 'hh:mm:ss');
end;

procedure TfrmWorldTime.Timer1Timer(Sender: TObject);
begin
  UpdateClocks;
end;

procedure TfrmWorldTime.FormCreate(Sender: TObject);
begin
  Application.Title := 'World Time';
  frmWorldTime.Constraints.MinHeight := LbDescription.Height + pnlClocks.Constraints.MinHeight +
    cxSplitter1.Height + pnlLocalTime.Constraints.MinHeight;
  frmWorldTime.Constraints.MinWidth := gcWashingtonTime.Constraints.MinWidth + Offset * gcWashingtonTime.Constraints.MinWidth;
  UpdateClocks;
end;

procedure TfrmWorldTime.FormShow(Sender: TObject);
begin
  Caption := Application.Title;
  Application.Hint := Caption;
end;

procedure TfrmWorldTime.pnlClocksResize(Sender: TObject);
begin
  pnlWashingtonClock.Width := (pnlClocks.Width - Offset * 4) div 3;
  pnlWashingtonClock.Height := pnlClocks.Height - Offset * 3;
  pnlWashingtonClock.Left := Offset;
  pnlWashingtonClock.Top := Offset;
  pnlParisClock.BoundsRect := cxRectOffsetHorz(pnlWashingtonClock.BoundsRect, pnlWashingtonClock.Width + Offset);
  pnlMoscowClock.BoundsRect := cxRectOffsetHorz(pnlParisClock.BoundsRect, pnlParisClock.Width + Offset);
end;

end.

File: Components/DevExpressVCL/Demos/ExpressGauge Control/CBuilder/WorldTime/uWorldTime.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uWorldTime.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxEdit"
#pragma link "cxLabel"
#pragma link "cxSplitter"
#pragma link "dxGaugeCircularScale"
#pragma link "dxGaugeControl"
#pragma link "dxGaugeCustomScale"
#pragma link "dxGaugeDigitalScale"
#pragma link "dxGaugeQuantitativeScale"
#pragma resource "*.dfm"
TfrmWorldTime *frmWorldTime;
int offset = 5;
//---------------------------------------------------------------------------
__fastcall TfrmWorldTime::TfrmWorldTime(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
TTimeInfo TfrmWorldTime::GetTimeInfo(int ATimeZone)
{
  TTimeInfo result;
  Word AMilli;
  TSystemTime ASystemTime;

  GetSystemTime(&ASystemTime);
  DecodeTime(SystemTimeToDateTime(ASystemTime), result.Hour, result.Min, result.Sec, AMilli);
  result.Hour = (result.Hour + ATimeZone) % 12;
  return result;
}
//---------------------------------------------------------------------------
void TfrmWorldTime::UpdateClock(TdxGaugeControl* AGaugeControl, TTimeInfo ATimeInfo)
{
  AGaugeControl->BeginUpdate();
  ((TdxGaugeCircularScale*)AGaugeControl->Scales->Items[3])->Value = ATimeInfo.Min;
  ((TdxGaugeCircularScale*)AGaugeControl->Scales->Items[2])->Value = ATimeInfo.Hour + ATimeInfo.Min / 60;
  ((TdxGaugeCircularScale*)AGaugeControl->Scales->Items[1])->Value = ATimeInfo.Sec;
  AGaugeControl->EndUpdate();
}
//---------------------------------------------------------------------------
void TfrmWorldTime::UpdateClocks()
{
  UpdateClock(gcWashingtonTime, GetTimeInfo(-5));
  UpdateClock(gcParisTime, GetTimeInfo(1));
  UpdateClock(gcMoscowTime, GetTimeInfo(3));
  gsDigital->Value = cxTimeToStr(Now(), "hh:mm:ss");
}
//---------------------------------------------------------------------------
void __fastcall TfrmWorldTime::FormCreate(TObject *Sender)
{
  Application->Title = "World Time";
  frmWorldTime->Constraints->MinHeight = lbDescription->Height + pnlClocks->Constraints->MinHeight +
	cxSplitter1->Height + pnlLocalTime->Constraints->MinHeight;
  frmWorldTime->Constraints->MinWidth = gcWashingtonTime->Constraints->MinWidth + offset * gcWashingtonTime->Constraints->MinWidth;
  CalculateGaugeBounds();
  UpdateClocks();
}
//---------------------------------------------------------------------------
void __fastcall TfrmWorldTime::Timer1Timer(TObject *Sender)
{
  UpdateClocks();
}
//---------------------------------------------------------------------------
void __fastcall TfrmWorldTime::FormShow(TObject *Sender)
{
  Caption = Application->Title;
  Application->Hint = Caption;
}
//---------------------------------------------------------------------------
void TfrmWorldTime::CalculateGaugeBounds()
{
  pnlWashingtonClock->Width = (pnlClocks->Width - offset * 4) / 3;
  pnlWashingtonClock->Height = pnlClocks->Height - offset * 3;
  pnlWashingtonClock->Left = offset;
  pnlWashingtonClock->Top = offset;
  pnlParisClock->BoundsRect = cxRectOffsetHorz(pnlWashingtonClock->BoundsRect, pnlWashingtonClock->Width + offset);
  pnlMoscowClock->BoundsRect = cxRectOffsetHorz(pnlParisClock->BoundsRect, pnlParisClock->Width + offset);
}
//---------------------------------------------------------------------------
void __fastcall TfrmWorldTime::pnlClocksResize(TObject *Sender)
{
  CalculateGaugeBounds();
}
//---------------------------------------------------------------------------

